packages/itwin/tree-widget/src/components/trees/stateless/models-tree/ModelsTreeDefinition.ts

```diff
@@ -226,22 +258,40 @@ export class ModelsTreeDefinition implements HierarchyDefinition {
                         WHEN (
                           json_extract([partition].JsonProperties, '$.PhysicalPartition.Model.Content') IS NOT NULL
                           OR json_extract([partition].JsonProperties, '$.GraphicalPartition3d.Model.Content') IS NOT NULL
+                          ${
+                            this._hierarchyConfig.showEmptyModels
+                              ? ""
+                              : `AND EXISTS (
+                                  SELECT 1
+                                  FROM ${this._hierarchyConfig.elementClassSpecification} element
+                                  WHERE element.Model.Id = m.ECInstanceId
+                                )`
+                          }
                         ) THEN 1
                         ELSE 0
                       END
                     `,
                   },
-                  hasChildren: true,
+                  hasChildren: {
+                    selector: `
+                      IFNULL((
+                        SELECT 1
+                        FROM ${this._hierarchyConfig.elementClassSpecification} e
+                        WHERE e.Model.Id = m.ECInstanceId
+                        LIMIT 1
+                      ), 0)
+                    `,
+                  },
```

Comment:
We should only use this selector if `showEmptyModels == true`. Otherwise, we know that the model has children, so we should just set this to `true`.

Reply:
Changed.